cpp/src/arrow/array/array_dict.cc

```diff
@@ -106,8 +106,9 @@ DictionaryArray::DictionaryArray(const std::shared_ptr<DataType>& type,
   SetData(data);
 }
 
-std::shared_ptr<Array> DictionaryArray::dictionary() const {
+const std::shared_ptr<Array>& DictionaryArray::dictionary() const {
   if (!dictionary_) {
+    // FIXME this isn't thread safe
```

Comment:
File issue and mention it here?

Reply:
Or move `dictionary_` init to constructor?

Reply:
Or use https://en.cppreference.com/w/cpp/memory/shared_ptr/atomic

Reply:
Atomic shared pointer tent to use `mutex` (folly has implement one using atomic swap). I guess it may hurts the performance

Reply:
Interesting, though an uncontended mutex should be reasonably fast.

Reply:
Follow up filed here https://github.com/apache/arrow/issues/36503

will update the comment to
```suggestion
    // TODO(GH-36503) this isn't thread safe
```